Sync v3-1-stable with v3-1-test to release 3.1.7rc1 (#61160)
mistercrunch/Airflow: RELEASE_NOTES.rst

.. towncrier release notes start

Airflow 3.1.7 (2026-02-03)
--------------------------

No significant changes.

Bug Fixes
^^^^^^^^^
- Fix asset scheduling for stale DAGs (#59337) (#60022) (#61106)
- Fix unnecessary DAG version churn when DAG file paths change (#60799)
- Fix missing warning when Bundle path may not be accessible to impersonated user (#60278)
- Fix ``TriggerDagRunOperator`` deferring when ``wait_for_completion=False`` (#60052)

> - Fix DAG processor OOM || Avoid loading all TaskInstances when checking DagVersion in write_dag (#60937)

Reviewer: ```suggestion
```

Duplicate - already listed below.

- Fix root logger to use log_level instead of hardcoded INFO level (#60784) (#60970)
- Fix DAG processor OOM || Avoid loading all TaskInstances when checking DagVersion in write_dag (#60937) (#60962)
- Fix worker startup DAG load failures by rescheduling tasks instead of exiting (#59604) (#60926)
- Fix permissions check in import error APIs (#60801) (#60884)
- Fix refresh-token invalidation by logging out the user (#60781) (#60881)
- Fix connection test API to restore masked password/extra from existing connections (#59643) (#60873)
- Fix dag-processor overhead by applying ``gc.freeze`` (#60505) (#60845)
- Fix dag-processor crashing due to MySql deadlock errors (#60166) (#60418)
- Unique run_id across manually triggered Dags with schedules (#59477) (#60468)
- UI: Plugins pagination fix  #61055 (#61059) (#61129)
- UI: Fix ui get dags permission endpoint for user without dag run permissions (#60979) (#60988)
- UI: Convert Tasks Table from card to table mode (#60830) (#60874)
- UI: Fix slow log scrolling for large task logs (#60806) (#60875)
- UI: Grey out trigger button on API 403 (#60648) (#60777)
- UI: Remove API error from disabling submit (#60473) (#60658)

> - UI: feat: added toasters for permission denied (#57966) (#58016) (#60646)

Reviewer: ```suggestion
- UI: Added toasters for permission denied (#57966) (#58016) (#60646)
```

- UI: Move row count and display toggle into DataTable (#57680) (#60639)
- UI: Fix/backfill permission error handling (#60582) (#60587)
- UI: Improve Dags Filter UI (#60346) (#60547)
- UI: Update PoolBar to separate Scheduled and Deferred slots (#59270) (#60538)

> - UI: Upgrade react-dom-router in Airflow UI (#60316) (#60456)

Reviewer: misc?

- UI: Correct the access for the ``externalLogUrl`` (#60412) (#60479)
- UI: Fix gantt chart styling (#60347) (#60457)

> - UI: Inherit core theme in React plugins (#60256) (#60405)

Reviewer: This doesn't feel bugfixy?

Reviewer: It was tagged for 3.2.0 too even :)

Author: it was backported by @pierrejeambrun https://github.com/apache/airflow/pull/60256#issuecomment-3737960883

Some of the UI changes are not fixes and hard to classify

Reviewer 2: Yes this could probably wait 3.2.0 it fixes an unreleased bug targeted for 3.2.0. Not sure why I back ported it, maybe boring cyborg auto backport label, and I did the backport without paying attention cause I usually put the tag myself.

> - UI: fix: reset pagination on DagList search (#60326) (#60336)

Reviewer: ```suggestion
- UI: Reset pagination on DagList search (#60326) (#60336)
```

- UI: Move dags list filters to ``buttongroups`` (#60298) (#60337)
- UI: Fix table filters resetting when deleting a dag (#60279) (#60287)
- UI: Fix sidebar visibility issue when main content exceeds ``viewport`` height (#59660) (#60286)
- UI: Add virtualization to grid view (#60241) (#60285)

misc/internal
"""""""""""""
- Use bulk DELETE for XComModel.clear() instead of loading records (#60955)
- Refactor DAG file queuing and fix redundant processing (#60124)
- Optimized DAG processing queue order on bundle refresh (#60003)
- Remove unused method ``is_default_pool`` in Pool model (#61084) (#61128)

Doc Only Changes
^^^^^^^^^^^^^^^^
- Fix minor display issue with migration to airflow 3 docs (#60749)
- Fix airflow.utils.context.Context import path in Airflow 3 migration doc (#59937)
- Add missing links to airflow.sdk classes and functions in public interface docs (#61005) (#61012)

> - Translation fixes: ``Taiwaness Mandarian`` (#61126), Catalan (#61093), German (#61097), Polish (#61099), Arabic (#60635 #60782),

Reviewer: These can't be doc-only right? Thats all in the ui.

Author: Yeah but were updating docs for UI. Should they be treated as bug fixes?

Reviewer: Misc is probably more appropriate.

  Spanish (#60775 #60785), Hebrew (#60633 #60686)

> - Docs: clarify BaseSensorOperator parameters in Sensors guide (#60275)

Reviewer: ```suggestion
- Clarify BaseSensorOperator parameters in Sensors guide (#60275)
```

.. towncrier release notes start

Airflow 3.1.7 (2026-02-04)
--------------------------

No significant changes.

Bug Fixes
^^^^^^^^^

> - Explicitly set table ordering (#60609) (#61216)

Reviewer: ```suggestion
- Explicitly set UI table ordering (#60609) (#61216)
```

> - Add guardrail to handle DAG deserialization errors in scheduler (#61162) (#61210)

Reviewer 3: ```suggestion
- Add guardrail to handle Dag deserialization errors in scheduler (#61162) (#61210)
```

- Avoid gantt annotation error during resize (#60877)
- Fix react apps plugins router (#61206)

> - Reset pagination on search in all pages (#61169)

Reviewer 3: If we wnat to group UI issues, should we add a `UI:` for these

- Fix asset scheduling for stale DAGs (#59337) (#60022) (#61106)

> - Fix unnecessary DAG version churn when DAG file paths change (#60799)

Reviewer 3: ```suggestion
- Fix asset scheduling for stale Dags (#59337) (#60022) (#61106)
- Fix unnecessary Dag version churn when Dag file paths change (#60799)
```

- Fix missing warning when Bundle path may not be accessible to impersonated user (#60278)
- Fix ``TriggerDagRunOperator`` deferring when ``wait_for_completion=False`` (#60052)

> - Fix NoneType error when updating serialized DAG (#56422)

Reviewer 3: ```suggestion
- Fix NoneType error when updating serialized Dag (#56422)
```

- Fix Pool API slots validation (#61071) (#61114)
- Fix DagBag parsing by adding bundle_path temporarily to sys.path (#55894) (#61053)
- Fix API to respect maximum page limit (#60989) (#61073)
- Prevent Triggerer from crashing when a trigger event isn't serializable (#60152) (#60981)
- Fix permissions on get_event_logs endpoint (#60936) (#60958)

> - Fix dag access control for dag_id in query param (#60935) (#60959)

Reviewer 3: ```suggestion
- Fix Dag access control for dag_id in query param (#60935) (#60959)
```

- Fix root logger to use log_level instead of hardcoded INFO level (#60784) (#60970)

> - Fix DAG processor OOM || Avoid loading all TaskInstances when checking DagVersion in write_dag (#60937) (#60962)

Reviewer 4: ```suggestion
- Fix DAG processor OOM by avoiding loading all TaskInstances when checking DagVersion in write_dag (#60937) (#60962)
```

> - Fix DAG processor OOM || Avoid loading all TaskInstances when checking DagVersion in write_dag (#60937) (#60962)

Reviewer 3: ```suggestion
- Fix Dag processor OOM by avoiding loading all TaskInstances when checking DagVersion in write_dag (#60937) (#60962)
```

> - Fix worker startup DAG load failures by rescheduling tasks instead of exiting (#59604) (#60926)

Reviewer 3: ```suggestion
- Fix worker startup Dag load failures by rescheduling tasks instead of exiting (#59604) (#60926)
```

- Fix permissions check in import error APIs (#60801) (#60884)
- Fix refresh-token invalidation by logging out the user (#60781) (#60881)
- Fix connection test API to restore masked password/extra from existing connections (#59643) (#60873)
- Fix dag-processor overhead by applying ``gc.freeze`` (#60505) (#60845)

> - Fix dag-processor crashing due to MySql deadlock errors (#60166) (#60418)

Reviewer 3: ```suggestion
- Fix Dag processor overhead by applying ``gc.freeze`` (#60505) (#60845)
- Fix Dag processor crashing due to MySQL deadlock errors (#60166) (#60418)
```

> - Unique run_id across manually triggered Dags with schedules (#59477) (#60468)

Reviewer 4: ```suggestion
- Ensure unique run_id across manually triggered Dags with schedules (#59477) (#60468)
```

- UI: Plugins pagination fix  #61055 (#61059) (#61129)

> - UI: Fix ui get dags permission endpoint for user without dag run permissions (#60979) (#60988)

Reviewer 3: ```suggestion
- UI: Fix UI get Dags permission endpoint for user without Dag run permissions (#60979) (#60988)
```

- UI: Convert Tasks Table from card to table mode (#60830) (#60874)
- UI: Fix slow log scrolling for large task logs (#60806) (#60875)
- UI: Grey out trigger button on API 403 (#60648) (#60777)
- UI: Remove API error from disabling submit (#60473) (#60658)
- UI: Added toasters for permission denied (#57966) (#58016) (#60646)
- UI: Move row count and display toggle into DataTable (#57680) (#60639)
- UI: Fix/backfill permission error handling (#60582) (#60587)
- UI: Improve Dags Filter UI (#60346) (#60547)
- UI: Update PoolBar to separate Scheduled and Deferred slots (#59270) (#60538)
- UI: Correct the access for the ``externalLogUrl`` (#60412) (#60479)
- UI: Fix gantt chart styling (#60347) (#60457)
- UI: Reset pagination on DagList search (#60326) (#60336)
- UI: Move dags list filters to ``buttongroups`` (#60298) (#60337)
- UI: Fix table filters resetting when deleting a dag (#60279) (#60287)
- UI: Fix sidebar visibility issue when main content exceeds ``viewport`` height (#59660) (#60286)
- UI: Add virtualization to grid view (#60241) (#60285)

> misc/internal

Reviewer 4: ```suggestion
Miscellaneous/Internal
```

> """""""""""""

Reviewer 4: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^
```

- UI: Upgrade react-dom-router in Airflow UI (#60316) (#60456)
- Use bulk DELETE for XComModel.clear() instead of loading records (#60955)
- Refactor DAG file queuing and fix redundant processing (#60124)

> - Optimized DAG processing queue order on bundle refresh (#60003)

Reviewer 3: ```suggestion
- Refactor Dag file queuing and fix redundant processing (#60124)
- Optimized Dag processing queue order on bundle refresh (#60003)
```

- Remove unused method ``is_default_pool`` in Pool model (#61084) (#61128)

Doc Only Changes
^^^^^^^^^^^^^^^^

> - Fix minor display issue with migration to airflow 3 docs (#60749)

Reviewer 3: ```suggestion
- Fix formatting issue with migration to Airflow 3 docs (#60749)
```

- Fix airflow.utils.context.Context import path in Airflow 3 migration doc (#59937)
- Add missing links to airflow.sdk classes and functions in public interface docs (#61005) (#61012)

> - Translation fixes: ``Taiwaness Mandarian`` (#61126), Catalan (#61093), German (#61097), Polish (#61099), Arabic (#60635 #60782),

Reviewer 3: ```suggestion
- Translation fixes: ``Taiwanese Mandarin`` (#61126), Catalan (#61093), German (#61097), Polish (#61099), Arabic (#60635 #60782),
```

Reviewer 3: my bad 🤦‍♂️
